clustermesh: export MCS-API data to clustermesh etcd

MrFreezeex opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [ ] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

This PR is about exporting MCS-API data to clustermesh etcd and do all the necessary changes including the integrationg the new struct to other clustermesh components like the kvstoremesh. 

Related to: #32712 #27902

```release-note
Export MCS-API data to clustermesh etcd to prepare for the ServiceImport auto creation
```

giorio94 commented on pkg/clustermesh/mcsapi/cell.go:
```diff
@@ -7,27 +7,34 @@ import (
+
+	cell.Provide(initMCSAPIController),
+	// Invoke an empty function which takes a mcsAPI to force its construction.
+	cell.Invoke(func(*mcsAPI) {}),
```
CRD to kvstore synchronization needs to be performed by either the clustermesh-apiserver or the Cilium operator, depending on whether Cilium is configured to operate in CRD or kvstore mode, respectively. You can find the clustermesh-apiserver core synchronization logic [here](https://github.com/cilium/cilium/blob/b30a3a941c54d5f74d0d1100936a74d340553476/clustermesh-apiserver/clustermesh/root.go#L377-L389). 

Moreover, The clustermesh-apiserver does not use controller runtime, but operates on `Resource[T]` directly. Unless strictly necessary, the synchronization of the ServiceExports should follow the same pattern, to avoid duplicating the services watcher. To that end, the `SyncStore` abstraction already leverages an internal working queue to handle retries, which don't need to be explicitly managed by the caller. You can take service synchronization as an [example](https://github.com/cilium/cilium/blob/b30a3a941c54d5f74d0d1100936a74d340553476/operator/watchers/k8s_service_sync.go#L92-L197).

Finally, in case KVStoreMesh is enabled, KVStoreMesh reads the data from the remote clusters and caches it into the local kvstore. You can find the main logic [here](https://github.com/cilium/cilium/blob/b30a3a941c54d5f74d0d1100936a74d340553476/pkg/clustermesh/kvstoremesh/remote_cluster.go), which needs to be adapted to additionally account for the newly introduced prefix.

MrFreezeex replied:
Ahh ok thanks, I didn't know that the sync could be executed into clustermesh-apiserver, I thought it was only the operator and I could thus use the sig-controller pattern. Will take a llok to change that then!

giorio94 commented on pkg/clustermesh/mcsapi/cell.go:
```diff
@@ -99,22 +106,58 @@ func checkRequiredCRDs(ctx context.Context, clientset k8sClient.Clientset) error
+	ctrlRuntimeManager                ctrl.Manager
+
+	factory        store.Factory
+	BackendPromise promise.Promise[kvstore.BackendOperations]
```
The operator does not currently provide the kvstore promise (due to historical reasons, as it would require further refactoring), and the client needs to be accessed via the global `kvstore.Client()` variable. The promise is instead provided by the clustermesh-apiserver.

giorio94 commented on pkg/clustermesh/mcsapi/mcsapiservicespec.go:
```diff
@@ -0,0 +1,118 @@
+}
+
+// Unmarshal parses the JSON byte slice and updates the MCS-API Service Spec receiver
+func (s *MCSAPIServiceSpec) Unmarshal(_ string, data []byte) error {
```
The unmarshal logic should perform basic validation to prevent processing incorrect data that may then cause a panic or unexpected behavior down the line. In this case, I'd suggest to validate that `Cluster`, `Namespace` and `Name` are not empty, that `Port` names are unique and `Port`  types are valid, and that the `Type` and `SessionAffinity` fields contain valid values. You can find an example for ClusterServices [here](https://github.com/cilium/cilium/blob/b30a3a941c54d5f74d0d1100936a74d340553476/pkg/service/store/store.go#L133-L164).

Additionally, when implementing the reader side, there should be additional validation to check that the cluster name matches the expected one, and that the kvstore key matches the namespace and name. Again, you can find an example for ClusterServices [here](https://github.com/cilium/cilium/blob/b30a3a941c54d5f74d0d1100936a74d340553476/pkg/service/store/store.go#L179-L223).

giorio94 reviewed (changes requested):
Thanks @MrFreezeex, I've left a couple of high level comments about the overall logic inline. Feel free to ask if you have doubts about them.

giorio94 commented on pkg/clustermesh/mcsapi/serviceexportsync_controller.go:
```diff
@@ -0,0 +1,199 @@
+				}
+				if !servicesSynced || !serviceExportsSynced {
+					continue
+				}
```
This doesn't look correct to me, because you'd miss all the events received before the initial synchronization (same for the one below).

MrFreezeex replied:
Ah yes indeed thanks for the catch, I wanted to queue the update actually

giorio94 commented on pkg/clustermesh/mcsapi/serviceexportsync_controller.go:
```diff
@@ -0,0 +1,199 @@
+				c.queue.Add(resource.Key{
+					Namespace: ev.Key.Namespace,
+					Name:      ev.Key.Name,
+				})
```
IMO, you should be able to further simplify this by completely getting rid of the working queue, and directly calling `syncMCSAPIServiceSpec` here. The reason being that retrieving the stores (that you perform above) blocks until the are fully synchronized. Hence, at this point you already have the complete data, and you can directly query the stores and propagate the information into the syncstore. WDYT?

MrFreezeex replied:
Ohh I didn't catch that getting a store was waiting for a full sync, then yes I suppose that we could do without a queue then :eyes:! The only thing is that with a queue I could de-duplicate updates but considering that ServiceExport has nothing past name and namespace it would probably only be somewhat useful when the service starts :thinking:.

giorio94 replied:
De-duplication is also already performed by the working queue of sync store, hence it would not add any additional benefit.

giorio94 commented on pkg/clustermesh/mcsapi/serviceexportsync_controller.go:
```diff
@@ -0,0 +1,199 @@
+	}
+}
+
+func (c *mcsAPIServiceSyncSpecSyncController) Run(ctx context.Context, serviceExportResource resource.Resource[*mcsapiv1alpha1.ServiceExport], wg *sync.WaitGroup, cfg watchers.ServiceSyncParameters) {
```
I'd suggest to make this method blocking rather than taking the `wg` as parameter, to simplify a bit the logic (it is also easier to propagate errors) and mimic the sync store behavior, but that's mostly a matter of preference.

giorio94 reviewed (commented):
Thanks! 

I've just looked at the `serviceexportsync_controller.go` file. The overall logic looks reasonable to me, I've left a couple of suggestions inline about possible simplifications. 

When preparing the final version of the PR, could you please break it into smaller commits for ease of review? :pray:

squeed commented on pkg/clustermesh/kvstoremesh/kvstoremesh.go:
```diff
@@ -165,6 +174,9 @@ func (km *KVStoreMesh) newRemoteCluster(name string, status common.StatusFunc) c
 
 	run(rc.nodes.syncer.Run)
 	run(rc.services.syncer.Run)
+	if km.mcsAPIConfig.ClusterMeshEnableMCSAPI {
```
Shouldn't this logic and code-base be entirely contained with in the `pkg/clustermesh/mcsapi` cell? This seems (at first glance) like a bit of a leak.

MrFreezeex replied:
Hmm not sure we can do much better than this tbh :/. This is all code coming from the rest of the `kvstoremesh` and is very similar to what we are doing for nodes, services and so on. The only thing that is specific to mcsapi is that we have a config to disable it but apart from that it's pretty much the same. However if you or anyone else have some suggestions/tips/leads to improve this I am happy to check/explore if it can be improved and do it if feasible!

giorio94 replied:
Hmm, I'd actually suggest to simplify this logic and always blindly attempt to synchronize this prefix in kvstoremesh. If not enabled, then nothing will need to be synchronized, so the end result would be the same in any case.

MrFreezeex replied:
I actually tried that initially but afaik the kvstoremesh was never ready if it was trying to sync service export while we were not exporting them, I didn't digged the exact reason though but I can check more

giorio94 replied:
I guess this is because the clustermesh-apiserver (or the operator) didn't write the "synced" flag for that resource type. One possibility could be just always writing it (i.e., starting the syncstore regardless of whether it is enabled, and calling `Synced`.

MrFreezeex replied:
I did that, it works locally and the code looks better thanks! There is one issue though, the ci-clustermesh upgrade tests are not happy :( (the job is this one: https://github.com/cilium/cilium/actions/runs/9669030476).

I am suspecting that it may be related to the fact that this is an upgrade test, so with one cluster not running this version of the code. What doesn't make sense though is that the failing task "waiting for cluster mesh status to be ready" says that the pod clustermesh-apiserver is started but not ready but in the pods dumps it is ready 😭. Eitherway I will try to reproduce locally this kind of upgrade scenario. If it is related to upgrade I am suspecting that it may be kvstoremesh trying to pull service exports from acluster that does not send the "synced" flag (because on an older version).

giorio94 replied:
> If it is related to upgrade I am suspecting that it may be kvstoremesh trying to pull service exports from acluster that does not send the "synced" flag (because on an older version).

Right, this is most likely the case, and kvstoremesh (as well as the agents, when that part will be implemented), need to tolerate the possibility that the serviceexport information is not yet present to support connecting to older versions. If you need to share whether serviceexports are present/enabled or not, then the [Capabilities section of the CiliumClusterConfig](https://github.com/cilium/cilium/blob/43642e5cda6eeb8c7ab7af9d64579ce725501e85/pkg/clustermesh/types/types.go#L80-L97) should be the right place for it.

MrFreezeex replied:
Many thanks for the extra pointer, so I implemented this change and could get a setup like this to work locally but the CI is still not happy with it :/. I am trying to dig why exactly...

MrFreezeex replied:
Ok so the CI issue seems be specifically related to the kvstoremesh commit (https://github.com/cilium/cilium/pull/32972/commits/bc0c2f7784969a1039bc5a9a52c80b88b3d9ccff).

I think I need to start the reflector (newReflector + run method on the syncer) only if the remote cluster support service exports. But... I don't have the cluster capability available in `newRemoteCluster`. So three different proposals to address this:
1) Bring back the flag `ClusterMeshEnableMCSAPI` in kstoremesh + remember to put a big warning in the doc to not activate mcs api support unless every clusters in the mesh has the cilium version necessary to run this (and perhaps keep the cluster capability but only as a mean to display an error log).

2) Have a way to add the service export reflector started in the `Run` method of the RemoteCluster

3) Refactor the RemoteCluster so that we can init all the reflectors at a time we have cluster capabilities available

WDYT @giorio94 :pray: ?

giorio94 replied:
From an initial look, I think that the missing part would be pretending that the synchronization completed in case service exports are disabled according to the capabilities (as it then eventually calls `o.synced.Done()`, which unblocks the readiness of kvstoremesh):

```diff
if srccfg.Capabilities.ServiceExportsSupport {
	serviceExportsSupported = true
	mgr.Register(adapter(mcsapitypes.ServiceExportStorePrefix), func(ctx context.Context) {
		rc.serviceExports.watcher.Watch(ctx, backend, path.Join(adapter(mcsapitypes.ServiceExportStorePrefix), rc.name))
	})
+} else {
+	rc.serviceExports.syncer.OnSync(ctx)
+}
```

Another possibility may be manually calling `o.synced.Done()` only, to not add the synced key in the kvstore (depending on which approach we want to adopt there). However, if that's the case, care must be payed to only call it once, and not for every reconnection. 

Additionally, it seems appropriate to me to call `rc.serviceExports.watcher.Drain()` as well, to ensure that cached entries are deleted upon reconnection if the `ServiceExportsSupport` value changes.

More in general, the reason why the reflectors are immediately started is that their lifecycle is independent of the connection to the kvstore itself, which can be performed multiple times. Additionally, the capabilities are not guaranteed to remain consistent (e.g., in case of upgrade), and possible changes of the capabilities need to be handled as well. 

(By the way, thanks for splitting the commits. IMO the PR is way easier to read now :rocket: )

MrFreezeex committed:
clustermesh: fix unrelated test name

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

MrFreezeex committed:
k8s: add mcsapi client to the clientset

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

MrFreezeex committed:
clustermesh: introduce the new kvstore mcsapi type

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

MrFreezeex committed:
clustermesh: add sync service export sync logic to kvstore

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

MrFreezeex committed:
api: add new clustermesh fields for MCS-API service exports

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

MrFreezeex committed:
operator: integrate MCS-API service exports sync start

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

MrFreezeex committed:
clustermesh-apiserver: integrate MCS-API service exports sync start

It also adds the necessary permission in the helm chart so that
clustermesh-apiserver is able to read local ServiceExport to be able to
perform the sync.

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

MrFreezeex committed:
clustermesh: add basic integration of mcsapi into clustermesh/operator

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

MrFreezeex committed:
clustermesh: allow kvstoremesh to replicate MCS-API service exports

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

MrFreezeex committed:
client: add MCS-API service exports support in the status

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

learnitall reviewed (commented):
There are some changes to the vendor directory, but I don't see any updates to `go.mod` or `go.sum`. Were these files missed?

MrFreezeex committed:
clustermesh: add service export enabled cluster capability

Add a new ServiceExportsEnabled field into CiliumClusterConfigCapabilities
to know what the MCS-API service exports feature status of the remote
cluster. This field is either nil (unsupported), true (enabled and
supported) or false (disabled and supported).

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

giorio94 commented on pkg/clustermesh/mcsapi/types/mcsapiservicespec.go:
```diff
@@ -0,0 +1,203 @@
+		return errors.New("name is unset")
+	case s.ExportCreationTimestamp.IsZero():
+		return errors.New("exportCreationTimestamp is unset")
+	case s.Type != mcsapiv1alpha1.ClusterSetIP && s.Type != mcsapiv1alpha1.Headless:
```
It seems reasonable to me to additionally validate that the `SessionAffinity` value is one of the supported two (or empty, if that's a valid possibility).

giorio94 commented on pkg/clustermesh/mcsapi/types/mcsapiservicespec.go:
I'd suggest adding a few unit tests to cover the logic introduced in this file. Mainly to ensure that serialization, deserialization and validation works as expected (instead of depending on the tests of the consumers).

giorio94 commented on pkg/clustermesh/mcsapi/types/mcsapiservicespec.go:
```diff
@@ -0,0 +1,203 @@
+
+// NewEmptyMCSAPIServiceSpec returns a MCSAPIServiceSpec with only the fields
+// needed to retrieve from the kvstore
+func NewEmptyMCSAPIServiceSpec(clusterName, name, namespace string) *MCSAPIServiceSpec {
```
Nit. the cluster, namespace, name order seems a bit more natural to me.

giorio94 commented on pkg/clustermesh/mcsapi/types/mcsapiservicespec.go:
```diff
@@ -0,0 +1,203 @@
+	ports := make([]mcsapiv1alpha1.ServicePort, 0, len(svc.Spec.Ports))
+	for _, port := range svc.Spec.Ports {
+		ports = append(ports, mcsapiv1alpha1.ServicePort{
+			Name:        port.Name,
```
The `Protocol` field seems to be missing.

giorio94 commented on pkg/clustermesh/mcsapi/serviceexportsync.go:
```diff
@@ -0,0 +1,154 @@
+	Services       resource.Resource[*slim_corev1.Service]
+	Backend        store.SyncStoreBackend
+	StoreFactory   store.Factory
+	Store          store.SyncStore
```
The `Store` parameter looks redundant to me.

MrFreezeex replied:
I did this so that I can create the store from the test to directly have a store client there instead of having to return it somehow but maybe you would have a better idea to do this :thinking:

giorio94 replied:
Fair enough. In that case, I'd unexport the `Store` field, and add a comment clarifying that it is used for testing purposes only. Another possibility could be modifying the test to not require mutating the SyncStore directly. Like first ensuring that the service export is present in the kvstore, then removing either the service or the service export and asserting that it gets removed. That would mimic even better the real behavior. Additionally, it would be better to use the fake client to create/remove the objects if possible, rather than mutating the underlying store, especially because it seems that the `CacheStore` method will be eventually dropped based on the comment.

giorio94 commented on pkg/clustermesh/mcsapi/serviceexportsync.go:
```diff
@@ -0,0 +1,154 @@
+}
+
+func StartSynchronizingServiceExports(ctx context.Context, wg *sync.WaitGroup, cfg ServiceExportSyncParameters) {
+	wg.Add(1)
+	go func() {
```
Nit. It could make sense to make this function blocking, to give more control to the caller (which can then run it in a goroutine if necessary).

giorio94 commented on pkg/clustermesh/mcsapi/serviceexportsync.go:
```diff
@@ -0,0 +1,154 @@
+		serviceEvents := cfg.Services.Events(ctx)
+		serviceStore, err := cfg.Services.Store(ctx)
+		if err != nil {
+			log.Errorf("can't init service store: %s", err)
```
Nit. The error should be specified with `.WithError`, rather than appending it to the message.

giorio94 commented on pkg/clustermesh/mcsapi/serviceexportsync_test.go:
```diff
@@ -0,0 +1,214 @@
+			},
+		},
+		{
+			TypeMeta: typeMetaSvcExport,
```
Nit. The `TypeMeta` doesn't look strictly necessary, but I'd be consistent, either always present or never present.

MrFreezeex replied:
Ah yes some forgotten copy paste from a test where it was necessary I believe :sweat_smile:.

giorio94 commented on pkg/clustermesh/mcsapi/serviceexportsync_test.go:
```diff
@@ -0,0 +1,214 @@
+		cell.Invoke(func(
+			svc resource.Resource[*slim_corev1.Service],
+			svcExport resource.Resource[*mcsapiv1alpha1.ServiceExport],
+		) error {
```
Nit. The error return parameter can be dropped.

giorio94 commented on clustermesh-apiserver/clustermesh/root.go:
```diff
@@ -360,8 +360,9 @@ func startServer(
-			MaxConnectedClusters: cinfo.MaxConnectedClusters,
+			SyncedCanaries:        true,
+			MaxConnectedClusters:  cinfo.MaxConnectedClusters,
+			ServiceExportsSupport: true,
```
I wonder if it could be actually useful to have this capability turned off in case service exports support is disabled. One reason may be to provide feedback on the client side when connecting to a cluster which has the feature disabled. WDYT?

giorio94 replied:
Additionally, the same capability should be also configured on the operator side (in case Cilium operates in kvstore mode): https://github.com/MrFreezeex/cilium/blob/svcexport-watcher-sync/operator/cmd/kvstore_watchdog.go#L109

MrFreezeex replied:
If it's for client reporting I believe we should somehow have a way to know if the feature is not supported, disabled (and supported) or enabled. As the sync to the kvstore mark as synced even if the feature is disabled we only need to do some special handling for a remote cluster in case the feature is not supported (and in case if later special handling is needed if the feature is disabled or enabled besides client reporting we would also have the info).

giorio94 commented on pkg/clustermesh/mcsapi/cell.go:
```diff
@@ -100,7 +86,7 @@ func checkRequiredCRDs(ctx context.Context, clientset k8sClient.Clientset) error
 
 func initMCSAPIController(params mcsAPIParams) error {
-	if !params.Clientset.IsEnabled() || params.ClusterMeshConfig == "" || !params.Cfg.ClusterMeshEnableMCSAPI {
+	if !params.Clientset.IsEnabled() || params.ClusterMeshConfig == "" || !params.CfgMCSAPI.ClusterMeshEnableMCSAPI {
```
Concerning `params.ClusterMeshConfig == ""`, it is actually safer checking whether `ClusterMesh != nil`, also for consistency with similar checks in the agent.

giorio94 commented on pkg/clustermesh/operator/clustermesh.go:
```diff
@@ -27,8 +27,10 @@ type clusterMesh struct {
-	logger  logrus.FieldLogger
-	Metrics Metrics
+	Cfg       ClusterMeshConfig
+	CfgMCSAPI MCSAPIConfig
```
Nit. Do these fields need to be exported?

MrFreezeex replied:
Nop, good catch! I lowercased them

giorio94 commented on pkg/clustermesh/operator/clustermesh.go:
```diff
@@ -63,18 +65,25 @@ type ClusterMesh interface {
 
 func newClusterMesh(lc cell.Lifecycle, params clusterMeshParams) (*clusterMesh, ClusterMesh) {
-	if params.ClusterMeshConfig == "" || !params.Cfg.ClusterMeshEnableEndpointSync {
+	if params.ClusterMeshConfig == "" {
```
Should additionally check that the ClusterID is not 0, for consistency with the same check on the agent side:

```go
if c.ClusterInfo.ID == 0 || c.ClusterMeshConfig == "" {
		return nil
}
```

giorio94 commented on pkg/clustermesh/operator/clustermesh.go:
```diff
@@ -137,6 +146,8 @@ func (cm *clusterMesh) GlobalServices() *common.GlobalServiceCache {
 	rc := &remoteCluster{
 		name:               name,
+		cfg:                cm.Cfg,
+		cfgMCSAPI:          cm.CfgMCSAPI,
```
Nit. How about propagating the individual values, rather than the entire struct? It would be likely easier to read.

giorio94 commented on pkg/clustermesh/operator/remote_cluster.go:
```diff
@@ -75,18 +97,25 @@ func (rc *remoteCluster) Remove() {
+	}
+	if rc.cfgMCSAPI.ClusterMeshEnableMCSAPI {
+		rc.remoteServiceExports.Drain()
+	}
```
Draining a watch store that has never been started is safe (and a no-op), hence you don't need the explicit check.

giorio94 commented on api/v1/openapi.yaml:
```diff
@@ -2533,6 +2533,9 @@ definitions:
       num-shared-services:
         description: Number of services in the cluster
         type: integer
+      num-service-exports:
+        description: Number of MCS-API service exports in the cluster
+        type: integer
```
It could make sense to additionally propagate the `ServiceExportsEnabled` capability in the appropriate section (especially if set to false when the feature is disabled).

giorio94 commented on pkg/clustermesh/operator/remote_cluster.go:
```diff
@@ -97,13 +126,32 @@ func (s *synced) Services(ctx context.Context) error {
+// cluster is disconnected, or the given context is canceled.
+func (s *synced) ServiceExports(ctx context.Context) error {
+	return s.Wait(ctx, s.serviceExports.WaitChannel())
+}
```
Depending on the expectations of the consumer, you may want to additionally mark the services as synced if disabled.

MrFreezeex replied:
Ahh I think I will just remove this for now, I didn't noticed it depends on code that I don't want to write in this PR (something a bit similar to GlobalServiceCache but for service exports/expanding the consumer interface)

giorio94 reviewed (commented):
I've started reviewing the initial commits (thanks again for splitting them). Overall the logic looks good to me, please find a few (mostly minor) comments inline.

giorio94 commented on pkg/clustermesh/mcsapi/types/mcsapiservicespec.go:
```diff
@@ -0,0 +1,215 @@
+	// Defaults to None.
+	// Ignored when type is Headless
+	// More info: https://kubernetes.io/docs/concepts/services-networking/service/#virtual-ips-and-service-proxies
+	SessionAffinity corev1.ServiceAffinity `json:"sessionAffinity,omitempty"`
```
Nit. Could it make sense to remove `omitempty` if we always expect it to be present?

giorio94 commented on pkg/clustermesh/mcsapi/types/mcsapiservicespec_test.go:
```diff
@@ -0,0 +1,183 @@
+
+func TestMCSAPIServiceSpec(t *testing.T) {
+	exportTime := metav1.Now().Rfc3339Copy()
+	exportTime.Time = exportTime.Local()
```
Any reason why this is required? It seems to imply that the test is somehow timezone dependent. Same for the other occurrences.

MrFreezeex replied:
Yep so that's because the marshal function of time from Kubernetes does this: https://github.com/cilium/cilium/blob/main/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/time.go#L136 and as we do not call the `Equal` method of the time object to compare but use the `require.Equal` we compare byte by byte and that's why we need it. If we would call the `Equal` method  from the time object we won't need it though

giorio94 replied:
Fair enough, could you please just add a comment in that case? I had tried removing it, and hit no issues on my side, hence why I was asking whether it was necessary.

MrFreezeex replied:
Yep will add a comment :+1:. FYI it also works on my laptop but didn't in the CI actually, I am not entirely sure what's different though but it means that adding this would very likely make it works on more stuff

giorio94 commented on pkg/clustermesh/mcsapi/types/mcsapiservicespec_test.go:
```diff
@@ -0,0 +1,183 @@
+	require.Equal(t, "cluster1/bar/foo", mcsAPISvcSpec.GetKeyName())
+
+	b, err := mcsAPISvcSpec.Marshal()
+	require.Nil(t, err)
```
Nit. `require.NoError` (same for all occurrences).

giorio94 commented on pkg/clustermesh/mcsapi/serviceexportsync.go:
```diff
@@ -0,0 +1,150 @@
+	SyncCallback   func(context.Context)
+}
+
+func StartSynchronizingServiceExports(ctx context.Context, wg *sync.WaitGroup, cfg ServiceExportSyncParameters) {
```
Nit. I wouldn't take the waitgroup as parameter, just create it in this function to ensure that when it returns all go routines terminated.

giorio94 commented on pkg/clustermesh/mcsapi/serviceexportsync.go:
```diff
@@ -0,0 +1,150 @@
+	if cfg.Store == nil {
+		if cfg.Backend == nil {
+			// Needs to be assigned in a separate goroutine, since it might block
+			// if the client is not yet initialized.
```
Nit. This comment doesn't look relevant anymore.

giorio94 commented on pkg/clustermesh/operator/remote_cluster.go:
```diff
@@ -97,13 +120,25 @@ func (s *synced) Services(ctx context.Context) error {
 
 	status.Synced = &models.RemoteClusterSynced{
-		Services: rc.remoteServices.Synced(),
+		Services:       rc.clusterMeshEnableEndpointSync && rc.remoteServices.Synced(),
```
Assuming that you want to mark it as synced if disabled, then the condition should be `!rc.clusterMeshEnableEndpointSync || rc.remoteServices.Synced()`.

giorio94 commented on pkg/clustermesh/operator/remote_cluster.go:
```diff
@@ -53,9 +65,19 @@ func (rc *remoteCluster) Run(ctx context.Context, backend kvstore.BackendOperati
+			rc.remoteServiceExports.Watch(ctx, backend, path.Join(adapter(mcsapitypes.ServiceExportStorePrefix), rc.name))
+		})
+	}
+	rc.serviceExportsEnabled = &serviceExportsEnabled
```
Access to this variable should be synchronized, as possibly accessed in parallel by multiple go routines. Additionally, it should be reset to nil when this function returns, as the status becomes unknown again at that point. 

Alternatively, considering that it is currently used for status reporting only, it could be dropped altogether, and replaced with something like:

```go
ServiceExports: !(rc.clusterMeshEnableMCSAPI && status.Config.ServiceExportsEnabled == nil || rc.isServiceExportsSynced(),
```

MrFreezeex replied:
Ahhhh I didn't noticed that there was a config available on the status object actually so yes I will do something like that, thanks!

giorio94 commented on pkg/clustermesh/operator/remote_cluster.go:
```diff
@@ -53,9 +65,19 @@ func (rc *remoteCluster) Run(ctx context.Context, backend kvstore.BackendOperati
+		serviceExportsEnabled = true
+		mgr.Register(adapter(mcsapitypes.ServiceExportStorePrefix), func(ctx context.Context) {
+			rc.remoteServiceExports.Watch(ctx, backend, path.Join(adapter(mcsapitypes.ServiceExportStorePrefix), rc.name))
+		})
```
You should call `rc.remoteServiceExports.Drain()` if `config.Capabilities.ServiceExportsEnabled == nil`, to ensure that any entries that could have been retrieved during a previous connection are flushed if service exports are no longer supported.

giorio94 commented on pkg/clustermesh/operator/remote_cluster.go:
This package appears to lack integration tests. I reckon that this PR is already on the very large size, and I'm OK with deferring them to a follow-up, but it would be great to ensure the correct functioning considering the different possible cases (i.e., different combinations of enabled/disabled/supported features). Another possibility could be moving the core of this commit to a separate PR, as it doesn't seem required for the subsequent changes (while you may need to still move the config definition to the different file here).

MrFreezeex replied:
Good idea I will move the changes from this file in another PR, I also realized with your other comment https://github.com/cilium/cilium/pull/32972#discussion_r1660672589 that I do need to have the service export cache to correctly init `remoteServiceExports` anyway which I didn't wanted to do here. So next PR would add a "service exports global cache" (or similar) and deals with remote_cluster.go + add some basic test for that as well

giorio94 replied:
:+1: (Feel free to assign me as reviewer on that PR as well, as I already have context on the changes)

giorio94 commented on pkg/clustermesh/operator/remote_cluster.go:
```diff
@@ -23,11 +24,22 @@ type remoteCluster struct {
 	remoteServices store.WatchStore
 
+	// remoteServiceExports is the shared store representing MCS-API service exports in remote clusters
+	remoteServiceExports store.WatchStore
```
`remoteServiceExports` does not appear to be ever initialized.

giorio94 commented on operator/cmd/kvstore_watchdog.go:
```diff
@@ -105,8 +105,11 @@ func startKvstoreWatchdog() {
+					ID: option.Config.ClusterID,
+					Capabilities: cmtypes.CiliumClusterConfigCapabilities{
+						MaxConnectedClusters:  option.Config.MaxConnectedClusters,
+						ServiceExportsEnabled: &option.Config.ClusterMeshEnableMCSAPI,
```
Given that the flag is registered through hive, it should be propagated here as a parameter, rather than adding a parallel field to the global struct.

MrFreezeex replied:
I looked at how to do this but it seems a bit weird to me as I would need to add the config in https://github.com/cilium/cilium/blob/main/operator/cmd/root.go#L468 and in this struct https://github.com/cilium/cilium/blob/main/operator/cmd/root.go#L491 and there is nothing else that look like a config there (or maybe there is another way to do this?) . Happy to do this if you think that's fine though :man_shrugging:.

giorio94 replied:
Yep, I know that it is suboptimal due to how the watchdog is started (it should a separate cell, but that would require a larger refactoring). But IMO it is still better than having a parallel global variable for the parameter.

MrFreezeex replied:
Ok sounds good thanks!

giorio94 commented on clustermesh-apiserver/clustermesh/k8s/resources.go:
```diff
@@ -24,8 +27,10 @@ var (
 		"Clustermesh-apiserver Kubernetes resources",
 
 		cell.Config(k8s.DefaultConfig),
+		cell.Config(operator.MCSAPIConfig{}),
```
Nit. It seems reasonable to me to move this to `clustermesh/cells.go`, as also used to populate the capabilities.

giorio94 replied:
The clustermesh-apiserver does not read the configuration from the configmap, and the flags should be propagated explicitly: https://github.com/MrFreezeex/cilium/blob/74a025c60a9467cb590fc28c848460e05e8513c6/install/kubernetes/cilium/templates/clustermesh-apiserver/deployment.yaml#L167-L215

giorio94 commented on pkg/client/client.go:
```diff
@@ -829,9 +829,9 @@ const (
 		if verbosity != RemoteClustersStatusNotReadyOnly || !cluster.Ready {
-			fmt.Fprintf(w, "   %s: %s, %d nodes, %d endpoints, %d identities, %d services, %d reconnections (last: %s)\n",
+			fmt.Fprintf(w, "   %s: %s, %d nodes, %d endpoints, %d identities, %d services, %d MCS-API service exports, %d reconnections (last: %s)\n",
 				cluster.Name, clusterReadiness(cluster), cluster.NumNodes,
-				cluster.NumEndpoints, cluster.NumIdentities, cluster.NumSharedServices,
+				cluster.NumEndpoints, cluster.NumIdentities, cluster.NumSharedServices, cluster.NumServiceExports,
```
The documentation entries displaying status snippets should be updated accordingly.

giorio94 commented on pkg/clustermesh/kvstoremesh/kvstoremesh_test.go:
```diff
@@ -111,75 +116,85 @@ func TestRemoteClusterRun(t *testing.T) {
 			srccfg: types.CiliumClusterConfig{
 				Capabilities: types.CiliumClusterConfigCapabilities{
-					SyncedCanaries: true,
+					ServiceExportsEnabled: ptrBool(true),
```
Nit. `ptr.To(true)`

giorio94 commented on pkg/clustermesh/kvstoremesh/remote_cluster.go:
```diff
@@ -101,6 +108,17 @@ func (rc *remoteCluster) Run(ctx context.Context, backend kvstore.BackendOperati
+	} else {
+		rc.serviceExports.syncer.OnSync(ctx)
+	}
+	rc.serviceExportsSupported = &serviceExportsSupported
```
Same comments here as for the logic in the operator.

giorio94 commented on pkg/clustermesh/kvstoremesh/remote_cluster.go:
```diff
@@ -101,6 +108,17 @@ func (rc *remoteCluster) Run(ctx context.Context, backend kvstore.BackendOperati
+			rc.serviceExports.watcher.Watch(ctx, backend, path.Join(adapter(mcsapitypes.ServiceExportStorePrefix), rc.name))
+		})
+	} else {
+		rc.serviceExports.syncer.OnSync(ctx)
```
Should make sure that this is called only at most once, and if not already previously synchronized (or make `OnSync` a no-op after the first execution). Currently, the watcher is guaranteed to trigger the sync callback (and in turn `OnSync` only once).

giorio94 commented on pkg/clustermesh/kvstoremesh/remote_cluster.go:
Now that https://github.com/cilium/cilium/pull/33153 got merged, ServiceExports should be additionally drained by the `drain` function.

giorio94 commented on clustermesh-apiserver/clustermesh/k8s/resources.go:
```diff
@@ -24,8 +27,10 @@ var (
+		cell.Config(operator.MCSAPIConfig{}),
 		cell.Provide(
 			k8s.ServiceResource,
+			mcsapi.ServiceExportResource,
```
Would it be appropriate to have a warning in case the MCS-API support is enabled, but the ServiceExports CRD is not present?

giorio94 reviewed (changes requested):
I've finished the first review pass, and left a bunch of additional comments inline. I'll do a full test once ready for review.

giorio94 commented on pkg/clustermesh/mcsapi/types/mcsapiservicespec.go:
```diff
@@ -0,0 +1,215 @@
+	case s.Type != mcsapiv1alpha1.ClusterSetIP && s.Type != mcsapiv1alpha1.Headless:
+		return fmt.Errorf("type is unknown: %s", s.Type)
+	case s.SessionAffinity != corev1.ServiceAffinityClientIP && s.SessionAffinity != corev1.ServiceAffinityNone:
+		return fmt.Errorf("session affinity is unknown: %s", s.Type)
```
`s/s.Type/s.SessionAffinity/`

giorio94 commented on pkg/clustermesh/mcsapi/serviceexportsync.go:
```diff
@@ -0,0 +1,157 @@
+	err := checkCRD(context.Background(), cs, mcsapiv1alpha1.SchemeGroupVersion.WithKind("serviceexports"))
+	if err != nil {
+		log.WithError(err).Warn("started Resource[ServiceExport] without the required CRD installed")
+	}
```
`checkCRD` cannot be executed from a `Provide` function, as they should not perform any interactions with external APIs. Indeed, the k8s client is not even supposed to be working yet at this point (although is actually does), as its [start hook](https://github.com/cilium/cilium/blob/4440433320c60fb35deda334913f3de7b3e7c659/pkg/k8s/client/cell.go#L236-L260) has not been executed yet.

Additionally, the clustermesh-apiserver doesn't currently have permissions to get CRDs (only list and watch :shrug:).
https://github.com/cilium/cilium/blob/4440433320c60fb35deda334913f3de7b3e7c659/install/kubernetes/cilium/templates/clustermesh-apiserver/clusterrole.yaml#L51-L57

MrFreezeex replied:
Makes sense, thanks for the explanation! I moved that to `StartSynchronizingServiceExports` and added the permission to get the crds

giorio94 commented on api/v1/openapi.yaml:
```diff
@@ -2591,6 +2597,10 @@ definitions:
         description: Whether the remote cluster supports per-prefix "synced" canaries
         type: boolean
+      service-exports-enabled:
+        description: Whether or not MCS-API ServiceExports is enabled by the cluster.
```
Nit. I'd add something like `... enabled by the cluster (null means unsupported).` for the sake of clarity.

giorio94 commented on pkg/clustermesh/operator/cell.go:
```diff
@@ -68,3 +70,18 @@ func (cfg ClusterMeshConfig) Flags(flags *pflag.FlagSet) {
+
+// ClusterMeshConfig contains the configuration for MCS-API
+type MCSAPIConfig struct {
+	// ClusterMeshEnableEndpointSync enables the MCS API support
```
Nit. `s/ClusterMeshEnableEndpointSync/ClusterMeshEnableMCSAPI/`. And two lines above `s/ClusterMeshConfig/MCSAPIConfig`.

giorio94 commented on pkg/clustermesh/operator/cell.go:
```diff
@@ -68,3 +70,18 @@ func (cfg ClusterMeshConfig) Flags(flags *pflag.FlagSet) {
+func (cfg MCSAPIConfig) Flags(flags *pflag.FlagSet) {
+	flags.BoolVar(&cfg.ClusterMeshEnableMCSAPI,
+		"clustermesh-enable-mcs-api",
+		false,
```
Nit. The default is typically provided as `cfg,ClusterMeshEnableMCSAPI`, so that it can be potentially configured when registering the configuration.

MrFreezeex replied:
I tried doing the same approach as what the kvstoremesh is doing, let me know if that's fine :+1:

giorio94 commented on pkg/clustermesh/operator/remote_cluster.go:
```diff
@@ -103,7 +108,8 @@ func (rc *remoteCluster) Status() *models.RemoteCluster {
 	status.Synced = &models.RemoteClusterSynced{
-		Services: rc.remoteServices.Synced(),
+		Services:       !rc.clusterMeshEnableEndpointSync || rc.remoteServices.Synced(),
+		ServiceExports: true, // FIXME
```
The comment says there's something to fix here.

MrFreezeex replied:
Yes it was a FIXME when the reader side would be integrated (not very explicit though indeed :sweat_smile:), now that I am changing this field to nullable I would drop the whole thing though

giorio94 commented on operator/cmd/kvstore_watchdog.go:
```diff
@@ -9,6 +9,7 @@ import (
 	"time"
 
 	"github.com/cilium/cilium/pkg/allocator"
+	operatorClusterMesh "github.com/cilium/cilium/pkg/clustermesh/operator"
```
Nit. `cmoperator` seems to align better to the aliases below.

giorio94 commented on pkg/clustermesh/mcsapi/serviceexportsync.go:
```diff
@@ -25,8 +26,8 @@ import (
-func ServiceExportResource(lc cell.Lifecycle, cs client.Clientset, opts ...func(*metav1.ListOptions)) resource.Resource[*mcsapiv1alpha1.ServiceExport] {
-	if !cs.IsEnabled() {
+func ServiceExportResource(lc cell.Lifecycle, cfg operator.MCSAPIConfig, cs client.Clientset, opts ...func(*metav1.ListOptions)) resource.Resource[*mcsapiv1alpha1.ServiceExport] {
+	if !cs.IsEnabled() || !cfg.ClusterMeshEnableMCSAPI {
```
I wonder whether it could be better to move this check to `StartSynchronizingServiceExports`, rather than embedding it here. The resource is lazily started only if any consumer calls `Events()` or `Store()` in any case.

MrFreezeex replied:
Ah yes if it's lazy I think it makes sense to do the prechecks in `StartSynchronizingServiceExports`  I deed :+1:

giorio94 commented on api/v1/openapi.yaml:
```diff
@@ -2564,6 +2567,9 @@ definitions:
       services:
         description: Services synchronization status
         type: boolean
+      service-exports:
```
Given that this is a new field, we may make it nullable to avoid needing to use the trick of setting it to true even if not watched. Then, you could likely drop https://github.com/cilium/cilium/pull/32972/commits/735de9a9401ef588bb10dc9754e6c23fb9461506.

giorio94 commented on pkg/clustermesh/kvstoremesh/remote_cluster.go:
```diff
@@ -286,7 +303,11 @@ func (o *syncer) OnDelete(key store.NamedKey) {
 func (o *syncer) OnSync(ctx context.Context) {
-	o.Synced(ctx, func(context.Context) { o.synced.Done() })
+	// As we send fake OnSync when service exports support is disabled we need
+	// to make sure that this is called onky once.
```
Nit. `s/onky/only`.

giorio94 commented on pkg/client/client.go:
```diff
@@ -843,9 +843,17 @@ func FormatStatusResponseRemoteClusters(w io.Writer, clusters []*models.RemoteCl
-					fmt.Fprintf(w, ", cluster-id=%d, kvstoremesh=%t, sync-canaries=%t",
-						cluster.Config.ClusterID, cluster.Config.Kvstoremesh, cluster.Config.SyncCanaries)
+					fmt.Fprintf(w, ", cluster-id=%d, kvstoremesh=%t, sync-canaries=%t, service-exports=%s",
+						cluster.Config.ClusterID, cluster.Config.Kvstoremesh, cluster.Config.SyncCanaries, serviceExportsConfig)
```
The synchronization status below should be updated to include service exports as well.

giorio94 commented on pkg/client/client.go:
```diff
@@ -843,9 +843,17 @@ func FormatStatusResponseRemoteClusters(w io.Writer, clusters []*models.RemoteCl
+					if *cluster.Config.ServiceExportsEnabled {
+						serviceExportsConfig = "disabled"
+					} else {
+						serviceExportsConfig = "enabled"
```
Disabled and enabled are inverted here.

giorio94 commented on pkg/clustermesh/operator/clustermesh.go:
```diff
@@ -67,14 +69,20 @@ type ClusterMesh interface {
+		return nil, nil
+	}
+
+	if !params.Cfg.ClusterMeshEnableEndpointSync && !params.CfgMCSAPI.ClusterMeshEnableMCSAPI {
```
This change should be reflected in the operator deployment as well:

https://github.com/cilium/cilium/blob/34055352b9253c99f64ae22ebce3e9de5ad6ab17/install/kubernetes/cilium/templates/cilium-operator/deployment.yaml#L95-L98

giorio94 reviewed (changes requested):
Overall looks great to me :rocket: Just a few last nits inline.

MrFreezeex commented on pkg/clustermesh/kvstoremesh/remote_cluster.go:
```diff
@@ -259,9 +274,13 @@ func (rc *remoteCluster) Status() *models.RemoteCluster {
 		Identities: rc.identities.watcher.Synced(),
 		Endpoints:  rc.ipcache.watcher.Synced(),
 	}
+	if status.Config != nil && status.Config.ServiceExportsEnabled != nil {
+		status.Synced.ServiceExports = ptr.To(rc.serviceExports.watcher.Synced())
+	}
```
I refactored a bit further this as well with the `ServiceExports` field that can now be nil, I think leaving this to nil unless we actually sync the service exports (remote cluster supports that) is more explicit and makes the logic slightly simpler.

giorio94 replied:
Makes sense to me.

giorio94 commented on pkg/clustermesh/mcsapi/serviceexportsync.go:
```diff
@@ -0,0 +1,164 @@
+		wg.Done()
+	}()
+
+	if !cfg.ClusterMeshEnableMCSAPI || cfg.ServiceExports == nil {
```
Nit. I'd add a comment to clarify why `Synced` is written even if the feature is disabled.

giorio94 commented on pkg/clustermesh/mcsapi/serviceexportsync.go:
```diff
@@ -0,0 +1,164 @@
+		return
+	}
+	if !cfg.skipCrdCheck {
+		err := checkCRD(context.Background(), cfg.Clientset, mcsapiv1alpha1.SchemeGroupVersion.WithKind("serviceexports"))
```
Nit. Could use `ctx`, rather than `context.Background()`.

giorio94 commented on pkg/clustermesh/kvstoremesh/remote_cluster.go:
```diff
@@ -108,6 +113,14 @@ func (rc *remoteCluster) Run(ctx context.Context, backend kvstore.BackendOperati
+			rc.serviceExports.watcher.Watch(ctx, backend, path.Join(adapter(mcsapitypes.ServiceExportStorePrefix), rc.name))
+		})
+	} else {
+		rc.serviceExports.syncer.OnSync(ctx)
```
I forgot to mention last time, but we should call `rc.serviceExports.watcher.Drain()` here as well, to ensure that any cached entry gets deleted in case we initially connected and the `ServiceExportsEnabled` capability was non-nil, then reconnected, and it is nil. Although I reckon this is pretty unlikely to happen, it is also harmless if no entries are present.

giorio94 commented on Documentation/operations/troubleshooting_clustermesh.rst:
```diff
@@ -74,10 +74,10 @@ you may perform the following steps to troubleshoot ClusterMesh issues.
-           └  remote configuration: expected=true, retrieved=true, cluster-id=3, kvstoremesh=false, sync-canaries=true
-           └  synchronization status: nodes=true, endpoints=true, identities=true, services=true
+           └  remote configuration: expected=true, retrieved=true, cluster-id=3, kvstoremesh=false, sync-canaries=true, service-exports=disabled
+           └  synchronization status: nodes=true, endpoints=true, identities=true, services=true, service-exports=nil
```
Nit. `nil` here could improved a bit, although I don't have any great idea. Maybe not showing the entry at all if nil?

MrFreezeex replied:
Good idea I went with your proposal of not showing the netry at all if nil :+1:

giorio94 commented on pkg/client/client.go:
```diff
@@ -843,18 +843,26 @@ func FormatStatusResponseRemoteClusters(w io.Writer, clusters []*models.RemoteCl
-				fmt.Fprintf(w, "   └  synchronization status: nodes=%v, endpoints=%v, identities=%v, services=%v\n",
-					cluster.Synced.Nodes, cluster.Synced.Endpoints, cluster.Synced.Identities, cluster.Synced.Services)
+				fmt.Fprintf(w, "   └  synchronization status: nodes=%v, endpoints=%v, identities=%v, services=%v, service-exports=%v\n",
+					cluster.Synced.Nodes, cluster.Synced.Endpoints, cluster.Synced.Identities, cluster.Synced.Services, cluster.Synced.ServiceExports)
```
The output doesn't look correct:

```
KVStoreMesh:    1/1 remote clusters ready
   clustermesh1: ready, 2 nodes, 8 endpoints, 3 identities, 0 services, 0 MCS-API service exports, 0 reconnections (last: never)
   └  etcd: 1/1 connected, leases=0, lock leases=1, has-quorum=true: endpoint status checks are disabled, ID: 80cdb2406fe49d9
   └  remote configuration: expected=true, retrieved=true, cluster-id=1, kvstoremesh=false, sync-canaries=true, service-exports=enabled
   └  synchronization status: nodes=true, endpoints=true, identities=true, services=true, service-exports=0xc00138d0aa
```

MrFreezeex replied:
Woops, indeed the `service-exports=0xc00138d0aa` should be fixed, was there anything else in the output that was incorrect?

giorio94 replied:
Nope, that was the only one I noticed.

learnitall reviewed (approved):
Vendor changes LGTM, thanks!

giorio94 reviewed (approved):
Looks good to me, thanks! Just a couple of final nits inline.

giorio94 reviewed (approved):
Thanks! It looks like you'll need to rebase to make CI happy.
